ModelProcessModelPlayerProxy should perform the work of SeparatedModelPlayer directly

eddydas opened the pull request:
#### 027113676fd2bb2657ee93ef93e5aff751129097
<pre>
ModelProcessModelPlayerProxy should perform the work of SeparatedModelPlayer directly
<a href="https://rdar.apple.com/124232154">rdar://124232154</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=270660">https://bugs.webkit.org/show_bug.cgi?id=270660</a>

Reviewed by Tim Horton.

ModelProcessModelPlayerProxy should should itself be a WebCore::ModelPlayer,
instead of further delegating it to SeparatedModelPlayer.

SeparatedModelPlayer was designed to work inside UIProcess, and thus has never
been an ideal choice for our purpose. By upstreaming the code from
SeparatedModelPlayer, we can now invoke the ModelPlayer calls directly through
XPC messages from ModelProcessModelPlayer without further boilerplate code
to propagate the calls.

At the same time we also upstream WKSeparatedModelLayer into the new
WKModelProcessModelLayer, which is a thin CALayer subclass just to listen to
layout and opacity changes should they occur.

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
Now conforms to WebCore::ModelPlayer properly, so that it can handle
all calls from ModelProcessModelPlayer 1:1.

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
Most code addition here are copied directly from SeparatedModelPlayer.
The rest are empty placeholders to satisfy WebCore::ModelPlayer
for the time being.

(WebKit::ModelProcessModelPlayerProxy::~ModelProcessModelPlayerProxy):
(WebKit::ModelProcessModelPlayerProxy::createLayer):
(WebKit::ModelProcessModelPlayerProxy::loadModel):
(WebKit::makeMeterSizeFromPointSize):
(WebKit::computeExtents):
(WebKit::degreesToRadians):
(WebKit::computeSRT):
(WebKit::effectivePointsPerMeter):
(WebKit::ModelProcessModelPlayerProxy::updateTransform):
(WebKit::ModelProcessModelPlayerProxy::updateOpacity):
(WebKit::ModelProcessModelPlayerProxy::startAnimating):
(WebKit::webDefaultLayerService):
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
(WebKit::ModelProcessModelPlayerProxy::didFailLoading):
(WebKit::ModelProcessModelPlayerProxy::load):
(WebKit::ModelProcessModelPlayerProxy::layer):
(WebKit::ModelProcessModelPlayerProxy::layerHostingContextIdentifier):
(WebKit::ModelProcessModelPlayerProxy::enterFullscreen):
(WebKit::ModelProcessModelPlayerProxy::supportsMouseInteraction):
(WebKit::ModelProcessModelPlayerProxy::supportsDragging):
(WebKit::ModelProcessModelPlayerProxy::setInteractionEnabled):
(WebKit::ModelProcessModelPlayerProxy::handleMouseDown):
(WebKit::ModelProcessModelPlayerProxy::handleMouseMove):
(WebKit::ModelProcessModelPlayerProxy::handleMouseUp):
(WebKit::ModelProcessModelPlayerProxy::getCamera):
(WebKit::ModelProcessModelPlayerProxy::setCamera):
(WebKit::ModelProcessModelPlayerProxy::isPlayingAnimation):
(WebKit::ModelProcessModelPlayerProxy::setAnimationIsPlaying):
(WebKit::ModelProcessModelPlayerProxy::isLoopingAnimation):
(WebKit::ModelProcessModelPlayerProxy::setIsLoopingAnimation):
(WebKit::ModelProcessModelPlayerProxy::animationDuration):
(WebKit::ModelProcessModelPlayerProxy::animationCurrentTime):
(WebKit::ModelProcessModelPlayerProxy::setAnimationCurrentTime):
(WebKit::ModelProcessModelPlayerProxy::hasAudio):
(WebKit::ModelProcessModelPlayerProxy::isMuted):
(WebKit::ModelProcessModelPlayerProxy::setIsMuted):
(WebKit::ModelProcessModelPlayerProxy::accessibilityChildren):
* Source/WebKit/ModelProcess/cocoa/WKModelProcessModelLayer.h: Added.
* Source/WebKit/ModelProcess/cocoa/WKModelProcessModelLayer.mm: Added.
(-[WKModelProcessModelLayer setPlayer:]):
(-[WKModelProcessModelLayer player]):
(-[WKModelProcessModelLayer setOpacity:]):
(-[WKModelProcessModelLayer layoutSublayers]):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/275940@main">https://commits.webkit.org/275940@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/76700ad394e66ead5a8cc312d59f66292385cfdb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43334 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22364 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45744 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45969 "Built successfully") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39460 "Hash 76700ad3 for PR 25613 does not build (failure)") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26114 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19782 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35821 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43908 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19384 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/37331 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/16801 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/16966 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1390 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/39545 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/38708 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47513 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/18248 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/15018 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/42609 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19785 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41266 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/19964 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5890 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19416 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
```diff
@@ -54,16 +65,46 @@ class ModelProcessModelPlayerProxy final
+    void didFailLoading(WebCore::REModelLoader&, const WebCore::ResourceError&) final;
+
+    // WebCore::ModelPlayer overrides.
+    void load(WebCore::Model&, WebCore::LayoutSize) override;
```
`s/override/final/` x100?

hortont424 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -90,10 +107,247 @@
+
+static CGFloat degreesToRadians(CGFloat degrees)
+{
+    return degrees * 3.14159 / 180;
```
M_PI?

hortont424 replied:
also I expect we have this code elsewhere

eddydas replied:
```
constexpr inline double deg2rad(double d)  { return d * radiansPerDegreeDouble; }
constexpr inline double rad2deg(double r)  { return r * degreesPerRadianDouble; }
```
in MathsExtras.h! Thanks for pointing out.

hortont424 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -90,10 +107,247 @@
+    [m_layer setValue:@YES forKeyPath:@"separatedOptions.updates.texture"];
+    [m_layer setValue:@YES forKeyPath:@"separatedOptions.updates.clippingPrimitive"];
+
+    // If I don't do this, I get black clearColor - why?!
```
Not this comment

hortont424 replied:
(or reword it if it needs to say)

eddydas replied:
Removed!

hortont424 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -90,10 +107,247 @@
+    m_loader = nullptr;
+
+    if (NSURL *url = error.nsError().userInfo[NSURLErrorFailingURLErrorKey])
+        RELEASE_LOG(ModelElement, "WebKit::SeparatedModelPlayer - Model failed to load with error \"%@\" (%@).", error.nsError().localizedDescription, url);
```
Definitely do not RELEASE_LOG URLs.

eddydas committed:
ModelProcessModelPlayerProxy should perform the work of SeparatedModelPlayer directly
rdar://124232154
https://bugs.webkit.org/show_bug.cgi?id=270660

Reviewed by Tim Horton.

ModelProcessModelPlayerProxy should should itself be a WebCore::ModelPlayer,
instead of further delegating it to SeparatedModelPlayer.

SeparatedModelPlayer was designed to work inside UIProcess, and thus has never
been an ideal choice for our purpose. By upstreaming the code from
SeparatedModelPlayer, we can now invoke the ModelPlayer calls directly through
XPC messages from ModelProcessModelPlayer without further boilerplate code
to propagate the calls.

At the same time we also upstream WKSeparatedModelLayer into the new
WKModelProcessModelLayer, which is a thin CALayer subclass just to listen to
layout and opacity changes should they occur.

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
Now conforms to WebCore::ModelPlayer properly, so that it can handle
all calls from ModelProcessModelPlayer 1:1.

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
Most code addition here are copied directly from SeparatedModelPlayer.
The rest are empty placeholders to satisfy WebCore::ModelPlayer
for the time being.

(WebKit::ModelProcessModelPlayerProxy::~ModelProcessModelPlayerProxy):
(WebKit::ModelProcessModelPlayerProxy::createLayer):
(WebKit::ModelProcessModelPlayerProxy::loadModel):
(WebKit::makeMeterSizeFromPointSize):
(WebKit::computeExtents):
(WebKit::degreesToRadians):
(WebKit::computeSRT):
(WebKit::effectivePointsPerMeter):
(WebKit::ModelProcessModelPlayerProxy::updateTransform):
(WebKit::ModelProcessModelPlayerProxy::updateOpacity):
(WebKit::ModelProcessModelPlayerProxy::startAnimating):
(WebKit::webDefaultLayerService):
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
(WebKit::ModelProcessModelPlayerProxy::didFailLoading):
(WebKit::ModelProcessModelPlayerProxy::load):
(WebKit::ModelProcessModelPlayerProxy::layer):
(WebKit::ModelProcessModelPlayerProxy::layerHostingContextIdentifier):
(WebKit::ModelProcessModelPlayerProxy::enterFullscreen):
(WebKit::ModelProcessModelPlayerProxy::supportsMouseInteraction):
(WebKit::ModelProcessModelPlayerProxy::supportsDragging):
(WebKit::ModelProcessModelPlayerProxy::setInteractionEnabled):
(WebKit::ModelProcessModelPlayerProxy::handleMouseDown):
(WebKit::ModelProcessModelPlayerProxy::handleMouseMove):
(WebKit::ModelProcessModelPlayerProxy::handleMouseUp):
(WebKit::ModelProcessModelPlayerProxy::getCamera):
(WebKit::ModelProcessModelPlayerProxy::setCamera):
(WebKit::ModelProcessModelPlayerProxy::isPlayingAnimation):
(WebKit::ModelProcessModelPlayerProxy::setAnimationIsPlaying):
(WebKit::ModelProcessModelPlayerProxy::isLoopingAnimation):
(WebKit::ModelProcessModelPlayerProxy::setIsLoopingAnimation):
(WebKit::ModelProcessModelPlayerProxy::animationDuration):
(WebKit::ModelProcessModelPlayerProxy::animationCurrentTime):
(WebKit::ModelProcessModelPlayerProxy::setAnimationCurrentTime):
(WebKit::ModelProcessModelPlayerProxy::hasAudio):
(WebKit::ModelProcessModelPlayerProxy::isMuted):
(WebKit::ModelProcessModelPlayerProxy::setIsMuted):
(WebKit::ModelProcessModelPlayerProxy::accessibilityChildren):
* Source/WebKit/ModelProcess/cocoa/WKModelProcessModelLayer.h: Added.
* Source/WebKit/ModelProcess/cocoa/WKModelProcessModelLayer.mm: Added.
(-[WKModelProcessModelLayer setPlayer:]):
(-[WKModelProcessModelLayer player]):
(-[WKModelProcessModelLayer setOpacity:]):
(-[WKModelProcessModelLayer layoutSublayers]):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/275940@main
